e2e/withdraw_test.go

```diff
@@ -0,0 +1,382 @@
+	testDoActionAndAssertTokenBalanceDifference(t, token, senderEthAddress, expectedBalanceDifference, func() {
+		txs := make([]types.Transaction, 0, depositsNeeded)
+		for i := 0; i < depositsNeeded; i++ {
+			tx, err := ethClient.QueueDeposit(queueDepositGasLimit, models.NewUint256(1), depositAmount, tokenID)
+			require.NoError(t, err)
+			txs = append(txs, *tx)
+		}
+		_, err := chain.WaitForMultipleTxs(ethClient.Blockchain.GetBackend(), txs...)
+		require.NoError(t, err)
+
+		waitForBatch(t, client, models.MakeUint256(1))
+	})
+}
+
+// userStatesDifference returns the user states in `a` that aren't in `b`.
+func userStatesDifference(a, b []dto.UserStateWithID) []dto.UserStateWithID {
+	mb := make(map[uint32]struct{}, len(b))
+	for _, x := range b {
+		mb[x.StateID] = struct{}{}
+	}
+	var diff []dto.UserStateWithID
+	for _, x := range a {
+		if _, found := mb[x.StateID]; !found {
+			diff = append(diff, x)
+		}
+	}
+	return diff
+}
+
+func testSendMassMigrationsForWithdrawal(
+	t *testing.T,
+	client jsonrpc.RPCClient,
+	senderWallet bls.Wallet,
+	fromStateID uint32,
+	amounts []models.Uint256,
+	startingNonce int,
+) common.Hash {
+	var firstTxHash common.Hash
+
+	for i := range amounts {
+		massMigration, err := api.SignMassMigration(&senderWallet, dto.MassMigration{
+			FromStateID: ref.Uint32(fromStateID),
+			SpokeID:     ref.Uint32(1),
+			Amount:      &amounts[i],
+			Fee:         models.NewUint256(1),
+			Nonce:       models.NewUint256(uint64(startingNonce + i)),
+		})
+		require.NoError(t, err)
+
+		var txHash common.Hash
+		err = client.CallFor(&txHash, "hubble_sendTransaction", []interface{}{*massMigration})
+		require.NoError(t, err)
+		require.NotZero(t, txHash)
+
+		testGetTransaction(t, client, txHash)
+
+		if i == 0 {
+			firstTxHash = txHash
+		}
+	}
+
+	return firstTxHash
+}
+
+func testSubmitWithdrawBatch(
+	t *testing.T,
+	client jsonrpc.RPCClient,
+	senderWallet bls.Wallet,
+	fromStateID uint32,
+) common.Hash {
+	massMigrationWithdrawalAmount := models.MakeUint256(uint64(9 * l2Unit))
+
+	var targetMassMigrationHash common.Hash
+	submitTxBatchAndWait(t, client, func() common.Hash {
+		targetMassMigrationHash = testSendMassMigrationsForWithdrawal(
+			t,
+			client,
+			senderWallet,
+			fromStateID,
+			[]models.Uint256{massMigrationWithdrawalAmount, models.MakeUint256(90)},
+			0,
+		)
+
+		return targetMassMigrationHash
+	})
+
+	return targetMassMigrationHash
+}
+
+func testGetMassMigrationCommitmentProof(
+	t *testing.T,
+	client jsonrpc.RPCClient,
+	batchID models.Uint256,
+	commitmentIndex uint8,
+) *dto.MassMigrationCommitmentProof {
+	var proof dto.MassMigrationCommitmentProof
+	err := client.CallFor(&proof, "hubble_getMassMigrationCommitmentProof", []interface{}{batchID, commitmentIndex})
+	require.NoError(t, err)
+
+	return &proof
+}
+
+func massMigrationCommitmentProofToGethType(proof *dto.MassMigrationCommitmentProof) withdrawmanager.TypesMMCommitmentInclusionProof {
```

Comment:
Maybe rename to `massMigrationCommitmentProofToCalldata` similar to functions that convert data for `submitBatch`